php/api/endpoint/class-report.php

```diff
@@ -0,0 +1,140 @@
+	 *
+	 * @action rest_api_init
+	 */
+	public function register_routes() {
+
+		register_rest_route( $this->namespace, '/' . $this->rest_base, array(
+			'methods'  => 'GET',
+			'callback' => array( $this, 'report_response' ),
+		) );
+
+		register_rest_route( $this->namespace, '/' . $this->rest_base . '/' . static::CHECKSUM_PATTERN . '/' . static::STANDARD_PATTERN, array(
+			'methods'  => 'GET',
+			'callback' => array( $this, 'report_response' ),
+		) );
+
+		register_rest_route( $this->namespace, '/' . $this->rest_base . '/' . static::POST_PATTERN . '/' . static::STANDARD_PATTERN, array(
+			'methods'  => 'GET',
+			'callback' => array( $this, 'report_response' ),
+		) );
+	}
+
+	/**
+	 * Prepare the response for a report requests.
+	 *
+	 * @param \WP_REST_Request $request The request object.
+	 *
+	 * @return mixed|\WP_REST_Response The response object.
+	 */
+	public function report_response( \WP_REST_Request $request ) {
+
+		// These reports are not for unauthenticated users.
+		if ( ! is_user_logged_in() ) {
+			return rest_ensure_response( $this->report_error( 'unauthenticated_call', __( 'unauthenticated report request', 'tide-api' ), 301 ) );
+		}
+
+		// @todo get the cloud storage source from configuration. For now this is AWS S3 only.
+		$object_source = 'aws_s3';
+
+		$checksum = $request->get_param( 'checksum' );
+		$post_id  = $request->get_param( 'post_id' );
+		$standard = $request->get_param( 'standard' );
+
+		// If we don't have a post_id, but a checksum, then hit the checksum endpoint.
+		if ( ! empty( $checksum ) ) {
+			$request  = new \WP_REST_Request( 'GET', '/' . $this->namespace . '/audit/' . $checksum );
+			$response = rest_do_request( $request );
+
+			if ( $response->is_error() ) {
+				// Don't be too verbose about the error.
+				return rest_ensure_response( $this->report_error( 'report_error', __( 'error occured in report api', 'tide-api' ), 301 ) );
+			}
+
+			// Set the post_id to get the meta for.
+			$post_id = $response->data['id'];
+		}
+
+		// Best effort to retrieve meta.
+		$meta = get_post_meta( $post_id, '_audit_' . $standard, true );
+
+		// Could not get the meta for the given standard.
+		if ( empty( $meta ) ) {
+			return rest_ensure_response( $this->report_error( 'report_standard_not_fount', __( 'could not retrieve report for standard', 'tide-api' ), 404 ) );
+		}
+
+		if ( empty( $this->plugin->components[ $object_source ] ) ) {
+			return rest_ensure_response( $this->report_error( 'report_source_error', __( 'could not retrieve report from source', 'tide-api' ), 404 ) );
+		}
+
+		// Get temporary signed url.
+		$meta = maybe_unserialize( $meta );
+		$url  = $this->plugin->components[ $object_source ]->get_url( $meta['full'] );
+
+		// Error fetching from S3.
+		if ( is_wp_error( $url ) ) {
+			return rest_ensure_response( $this->report_error( 'report_fetch_error', __( 'fetching report failed', 'tide-api' ), 500 ) );
+		}
+
+		// Redirect to URL to download.
+		wp_redirect( $url );
```

Comment:
Are we sure we should be redirecting?

Reply:
If we want the browser to start downloading on click... then yes. The alternative is to present a separate page with a "Click Here to Download" link.   Try it out with a curl request or with Postman and you will see we get the desired result.

Reply:
Another alternative is to do what we did before.... use PHP to "download" the object and echo it out. It feels nasty and slow.  This way a click on the _link will immediately download the file.

Reply:
I'm thinking about the use case of a front-end UI that want's to get the linked resource. If they follow the link to inject that data into a component on the page asynchronously will a redirect still work?

Reply:
I'll have to whip up a quick JS test to confirm.  If postman and curl does it, I wouldn't see why not... but let me check.

Reply:
As suspected... JS did give me a CORS issue.  I will patch this with our other suggested solution.